File: 3331-find-subtree-sizes-after-changes/3331-find-subtree-sizes-after-changes.cpp
class Solution {
public:
    vector<int> findSubtreeSizes(vector<int>& parent, string s) {
        int n = parent.size();
        vector<vector<int>> tree(n);
        for(int i=1;i<n;i++) tree[parent[i]].push_back(i);
        
        vector<int> new_p(parent);
        vector<int> last(26, -1);
        struct Frame {
            int node;
            bool is_exit;
            int old_val;
        };
        vector<Frame> stack_frames;
        stack_frames.push_back(Frame{0, false, -1});
        
        while(!stack_frames.empty()){
            Frame f = stack_frames.back(); stack_frames.pop_back();
            if(!f.is_exit){
                int c = s[f.node]-'a';
                int old = last[c];
                if(old != -1) new_p[f.node] = old;
                stack_frames.push_back(Frame{f.node, true, last[c]});
                last[c] = f.node;
                for(auto it = tree[f.node].rbegin(); it!=tree[f.node].rend(); it++) stack_frames.push_back(Frame{*it, false, -1});
            }
            else{
                int c = s[f.node]-'a';
                last[c] = f.old_val;
            }
        }
        
        vector<vector<int>> new_tree(n);
        for(int i=0;i<n;i++) if(new_p[i]!=-1) new_tree[new_p[i]].push_back(i);
        
        vector<int> size_sub(n, 1);
        stack_frames.push_back(Frame{0, false, -1});
        while(!stack_frames.empty()){
            Frame f = stack_frames.back(); stack_frames.pop_back();
            if(!f.is_exit){
                stack_frames.push_back(Frame{f.node, true, -1});
                for(auto it = new_tree[f.node].rbegin(); it!=new_tree[f.node].rend(); it++) stack_frames.push_back(Frame{*it, false, -1});
            }
            else{
                for(auto child: new_tree[f.node]) size_sub[f.node] += size_sub[child];
            }
        }
        return size_sub;
    }
};

File: 3331-find-subtree-sizes-after-changes/3331-find-subtree-sizes-after-changes.java
class Solution {
    private static class Frame {
         int node;
         boolean isExit;
         int prevLastSeen;

         Frame(int node, boolean isExit, int prevLastSeen) {
             this.node = node;
             this.isExit = isExit;
             this.prevLastSeen = prevLastSeen;
         }
     }

     public int[] findSubtreeSizes(int[] parent, String s) {
         int n = parent.length;

         List<List<Integer>> tree = new ArrayList<>(n);
         for (int i = 0; i < n; i++) {
             tree.add(new ArrayList<>());
         }
         for (int i = 1; i < n; i++) {
             tree.get(parent[i]).add(i);
         }

         int[] newParent = new int[n];
         System.arraycopy(parent, 0, newParent, 0, n);

         int[] lastSeen = new int[26];
         Arrays.fill(lastSeen, -1);

         Deque<Frame> stack = new ArrayDeque<>();
         stack.push(new Frame(0, false, -1)); 

         while (!stack.isEmpty()) {
             Frame current = stack.pop();
             int node = current.node;

             if (!current.isExit) {
                 char c = s.charAt(node);
                 int cIdx = c - 'a';
                 int y = lastSeen[cIdx];

                 if (y != -1) {
                     newParent[node] = y;
                 }

                 
                 stack.push(new Frame(node, true, lastSeen[cIdx]));

                
                 lastSeen[cIdx] = node;

                 
                 List<Integer> children = tree.get(node);
                 for (int i = children.size() - 1; i >= 0; i--) {
                     stack.push(new Frame(children.get(i), false, -1));
                 }
             } else {
                 
                 char c = s.charAt(node);
                 int cIdx = c - 'a';
                 lastSeen[cIdx] = current.prevLastSeen;
             }
         }


         List<List<Integer>> newTree = new ArrayList<>(n);
         for (int i = 0; i < n; i++) {
             newTree.add(new ArrayList<>());
         }
         for (int i = 1; i < n; i++) {
             newTree.get(newParent[i]).add(i);
         }

       
         int[] size = new int[n];
         stack.push(new Frame(0, false, -1)); // Start with root node

         while (!stack.isEmpty()) {
             Frame current = stack.pop();
             int node = current.node;

             if (!current.isExit) {
                 
                 stack.push(new Frame(node, true, -1));

               
                 List<Integer> children = newTree.get(node);
                 for (int i = children.size() - 1; i >= 0; i--) {
                     stack.push(new Frame(children.get(i), false, -1));
                 }
             } else {
                
                 size[node] = 1; // Count the node itself
                 for (int child : newTree.get(node)) {
                     size[node] += size[child];
                 }
             }
         }

         return size;
    }
}